File: zkm_cloud_test/ha-print.py
#!/usr/bin/env python3
"""Print a Home Assistant sensor as plain lines: time + value.

Same options and endpoints as marsSensorData.py, standard library only.
Nothing to do with the station or its container — talks to Home Assistant.

  python3 ha-print.py --env path/to/.env                 current value of every station sensor
  python3 ha-print.py --env path/to/.env --day           every reading of the last 24 h
  python3 ha-print.py --env path/to/.env --hour          the last hour
  python3 ha-print.py --env path/to/.env -s shelly_steckdose_2_energie --day
  HA_HOST=192.168.225.23 HA_API_TOKEN=... python3 ha-print.py --day     (without a .env)
  python3 ha-print.py ... --json                          raw JSON as Home Assistant sends it
"""

import json
import os
import sys
import urllib.error
import urllib.request
from datetime import datetime, timedelta, timezone
from zoneinfo import ZoneInfo

TZ = ZoneInfo("Europe/Berlin")

# The station's sensors (content/home-assistant.json) — entity ids without "sensor."
SENSORS = [
    ("sonoff_temperatur_temperature_1", "Cricket Terrarium Temperature"),
    ("shelly_steckdose_2_energie", "Socket energy · Shelly plug"),
]

HA_HOST = os.environ.get("HA_HOST", "")
HA_PORT = os.environ.get("HA_PORT", "80")
HA_API_TOKEN = os.environ.get("HA_API_TOKEN", "")

MODE = "now"       # now | day | hour
SENSOR = None      # one entity, or None for all in SENSORS
RAW = False
ENV_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), ".env")


def load_env(path):
    """Read HA_HOST / HA_PORT / HA_API_TOKEN from the station's .env (never printed)."""
    global HA_HOST, HA_PORT, HA_API_TOKEN
    try:
        with open(path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                k, v = line.split("=", 1)
                v = v.strip().strip('"').strip("'")
                if k.strip() == "HA_HOST" and not os.environ.get("HA_HOST"):
                    HA_HOST = v
                elif k.strip() == "HA_PORT" and not os.environ.get("HA_PORT"):
                    HA_PORT = v
                elif k.strip() == "HA_API_TOKEN" and not os.environ.get("HA_API_TOKEN"):
                    HA_API_TOKEN = v
        return True
    except OSError:
        return False


args = sys.argv[1:]
i = 0
while i < len(args):
    key = args[i]
    if key in ("--help", "help"):
        print(__doc__)
        sys.exit(0)
    elif key in ("-h", "--host"):
        HA_HOST = args[i + 1]; i += 2
    elif key in ("-p", "--port"):
        HA_PORT = args[i + 1]; i += 2
    elif key in ("-s", "--sensor"):
        SENSOR = args[i + 1].removeprefix("sensor."); i += 2
    elif key in ("-t", "--token"):
        HA_API_TOKEN = args[i + 1]; i += 2
    elif key == "--env":
        ENV_FILE = args[i + 1]; i += 2
    elif key == "--hour":
        MODE = "hour"; i += 1
    elif key in ("-d", "--day"):
        MODE = "day"; i += 1
    elif key in ("-n", "--now"):
        MODE = "now"; i += 1
    elif key == "--json":
        RAW = True; i += 1
    else:
        print(f"ERROR: unknown argument {key}", file=sys.stderr)
        sys.exit(1)

env_loaded = load_env(ENV_FILE)

if not HA_API_TOKEN:
    print("ERROR: no API token — give --env path/to/mars-station/.env, or --token TOKEN, or HA_API_TOKEN=...", file=sys.stderr)
    sys.exit(1)
if not HA_HOST:
    print("ERROR: no host — give --env path/to/mars-station/.env, or --host HOST, or HA_HOST=...", file=sys.stderr)
    sys.exit(1)


def call(url):
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {HA_API_TOKEN}", "Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=10) as r:
            return r.status, r.read().decode("utf-8", "replace").strip()
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8", "replace").strip()
    except urllib.error.URLError as e:
        print(f"ERROR: cannot reach {url}\n  {e.reason}\n  This PC must be on the same network as the habitat hardware.", file=sys.stderr)
        sys.exit(1)


def when(iso):
    if not iso:
        return "—"
    return datetime.fromisoformat(iso.replace("Z", "+00:00")).astimezone(TZ).strftime("%d/%m/%Y %H:%M:%S")


def fetch(sensor):
    base = f"http://{HA_HOST}:{HA_PORT}"
    if MODE == "now":
        return call(f"{base}/api/states/sensor.{sensor}")
    period = ""
    if MODE == "hour":
        period = "/" + (datetime.now(timezone.utc) - timedelta(hours=1)).strftime("%Y-%m-%dT%H:%M:%S+00:00")
    return call(f"{base}/api/history/period{period}?filter_entity_id=sensor.{sensor}&minimal_response")


print(f"Home Assistant at http://{HA_HOST}:{HA_PORT}  ({'settings from ' + ENV_FILE if env_loaded else 'settings from environment / arguments'})")
print()

targets = [(SENSOR, SENSOR)] if SENSOR else SENSORS
for sensor, label in targets:
    status, body = fetch(sensor)
    if status != 200:
        print(f"{label}  [sensor.{sensor}]\n  HTTP {status}: {body[:200]}\n")
        continue
    data = json.loads(body)
    if RAW:
        print(json.dumps(data, indent=2, ensure_ascii=False))
        print()
        continue

    if MODE == "now":
        unit = (data.get("attributes") or {}).get("unit_of_measurement", "")
        name = (data.get("attributes") or {}).get("friendly_name") or label
        print(f"{name}  [sensor.{sensor}]")
        print(f"  now: {data.get('state')} {unit}  at {when(data.get('last_updated'))}")
    else:
        rows = data[0] if data else []
        # with minimal_response only the first row carries attributes
        unit = ((rows[0].get("attributes") or {}) if rows else {}).get("unit_of_measurement", "")
        name = ((rows[0].get("attributes") or {}) if rows else {}).get("friendly_name") or label
        print(f"{name}  [sensor.{sensor}]  — {'last hour' if MODE == 'hour' else 'last 24 h'}, {len(rows)} readings")
        if not rows:
            print("  (nothing recorded in this window)")
        for r in rows:
            state = r.get("state")
            if state in ("unavailable", "unknown", None):
                continue
            print(f"  {when(r.get('last_changed') or r.get('last_updated'))}   {state} {unit}")
    print()

File: Claude outputs/ha-print.js
#!/usr/bin/env node
// Calls the Home Assistant API directly and prints the sensors. Nothing to do
// with the station or its container — talks to the hardware's own server.
//
//   node ha-print.js                     current value of each sensor
//   node ha-print.js --history           every reading of the last 24 h, time + value
//   node ha-print.js --history --hours 6
//   node ha-print.js --watch             re-read every 10 s until Ctrl-C
//
// Where Home Assistant is and the token: the script reads HA_HOST / HA_PORT /
// HA_API_TOKEN from the station's .env — give its path with --env, or copy
// the .env next to this script. Values below are the fallback.
// Needs only Node 18 or newer.

const fs = require('fs');
const path = require('path');
const envFile = (() => { const i = process.argv.indexOf('--env'); if (i > 0) { const p = process.argv[i + 1]; process.argv.splice(i, 2); return p; } return path.join(__dirname, '.env'); })();
const ENV = {};
try {
  for (const line of fs.readFileSync(envFile, 'utf8').split(/\r?\n/)) {
    const m = line.match(/^\s*([A-Z0-9_]+)\s*=\s*(.*?)\s*$/); if (m) ENV[m[1]] = m[2].replace(/^["']|["']$/g, '');
  }
} catch { /* no .env here — fall back to the values below */ }

const HOST = process.env.HA_HOST || ENV.HA_HOST || '192.168.225.23';   // Home Assistant's IP (from the station's .env)
const PORT = process.env.HA_PORT || ENV.HA_PORT || '80';
const TOKEN = process.env.HA_API_TOKEN || ENV.HA_API_TOKEN || '';      // long-lived token — never write it into this file

// Entity ids to print — the same ones the station shows.
const SENSORS = [
  'sensor.sonoff_temperatur_temperature_1',   // Cricket Terrarium Temperature
  'sensor.shelly_steckdose_2_energie',        // Socket energy · Shelly plug
];

const args = process.argv.slice(2);
const flag = (n) => { const i = args.indexOf(n); if (i < 0) return false; args.splice(i, 1); return true; };
const opt = (n, d) => { const i = args.indexOf(n); if (i < 0) return d; const v = args[i + 1]; args.splice(i, 2); return v; };
const history = flag('--history');
const watch = flag('--watch');
const hours = Number(opt('--hours', 24)) || 24;

const base = `http://${HOST}:${PORT}`;
const when = (iso) => new Date(iso).toLocaleString('en-GB', { timeZone: 'Europe/Berlin', hour12: false }).replace(',', '');

async function ha(path) {
  const res = await fetch(base + path, { headers: { Authorization: `Bearer ${TOKEN}` } });
  if (res.status === 401) throw new Error('401 — the token is wrong or expired');
  if (res.status === 404) throw new Error('404 — no such entity');
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function once() {
  if (watch) console.clear();
  for (const id of SENSORS) {
    try {
      const s = await ha(`/api/states/${id}`);
      const unit = s.attributes?.unit_of_measurement || '';
      console.log(`${s.attributes?.friendly_name || id}  [${id}]`);
      console.log(`  now: ${s.state} ${unit}  at ${when(s.last_updated)}`);
      if (history) {
        const start = new Date(Date.now() - hours * 3600000).toISOString();
        const rows = (await ha(`/api/history/period/${start}?filter_entity_id=${id}&minimal_response&no_attributes`))[0] || [];
        for (const r of rows) {
          if (r.state === 'unavailable' || r.state === 'unknown') continue;
          console.log(`  ${when(r.last_changed || r.last_updated)}   ${r.state} ${unit}`);
        }
        if (!rows.length) console.log(`  (no history in the last ${hours} h)`);
      }
    } catch (e) {
      console.log(`${id}\n  could not read — ${e.message}`);
    }
    console.log('');
  }
}

(async () => {
  if (!TOKEN) { console.error(`no HA_API_TOKEN found (looked in ${envFile}). Run:  node ha-print.js --env "C:\\path\\to\\mars-station\\.env"`); process.exit(1); }
  console.log(`Home Assistant at ${base}  (settings from ${fs.existsSync(envFile) ? envFile : 'built-in fallback'})\n`);
  try { await fetch(base + '/api/', { headers: { Authorization: `Bearer ${TOKEN}` } }); }
  catch (e) { console.error(`cannot reach Home Assistant at ${base} — ${e.message}\nThis PC must be on the same network as the habitat hardware (the station's container reaches it from there).`); process.exit(1); }
  await once();
  if (watch) setInterval(once, 10_000);
})();
